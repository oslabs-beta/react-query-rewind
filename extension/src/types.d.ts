// component prop types

export type TabPanelProps = {
  children?: React.ReactNode;
  index: number;
  value: number;
};

export type BasicTabsProps = {
  selectedQueries: string[];
  queryEvents: QueryEvent[];
};

export type ParentTabsProps = {
  // queryOptions: string[];
  queryEvents: QueryEvent[];
  selectedQueries: string[];
  handleSelectionChange: (queries: string[]) => void;
};

export type QueryTabProps = {
  selectedQueries: string[];
  queryEvents: QueryEvent[];
  handleSelectionChange: (queries: string[]) => void;
};

export type DataTabProps = {
  queryDisplay: QueryDisplay[][];
  currentIndex: number;
};

export type SliderSectionProps = {
  queryDisplay: QueryDisplay[][];
  currentIndex: number;
  setCurrentIndex: (selected: number) => void;
  handleAutoPlay: () => void;
  selectedQueries: string[];
  isPlaying: boolean;
};

export type JsonFormatterType = {
  jsonData: JsonDataType;
  queryKey: string;
  expandNodesFunc?: ExpandNodesFuncType; // this should be required but getting errors when it's not
};

export type JsonDiffType = {
  oldJson?: JsonDataType | string, // optional in case you're on first state
  currentJson: JsonDataType | string, // or string since state gets initialized to an empty string
  queryKey: string,
  isHidden: boolean
};

// variable types
export type QueryEvent = {
  eventType: string;
  queryKey: string[];
  queryHash: string;
  timestamp: Date;
  queryData?: any;
};

export type QueryData = {
  [queryName: string]: {
    updates: QueryEvent[];
  };
};

export type QueryDisplay = {
  queryKey: string;
  queryData: any;
};

export type JsonDataType = {
  [key: string]: any;
};

// function types

export type ExpandNodesFuncType = (
  keyPath: ReadonlyArray<string | number>,
  value: any,
  layer: number
) => boolean