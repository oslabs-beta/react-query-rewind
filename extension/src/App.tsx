import './css/styles.css';
import React, { useState, useEffect } from 'react';
import ParentTab from './containers/ParentTab';
import { QueryEvent } from './types';
import MultiSelect from './components/MultiSelect';
import saveSelectedQueryKeys from './functions/saveSelectedQueryKeys'
import Box from '@mui/material/Box';
import Container from '@mui/material/Container';
import { createTheme } from '@mui/material/styles';

type QueryMetrics = {
  // [queryKey: ]
};

function App() {
  // state to store changes to query cache
  const [queryEvents, setQueryEvents] = useState<QueryEvent[]>([]);
  const [queryMetrics, setQueryMetrics] = useState();
  const [selectedQueries, setSelectedQueries] = useState<string[]>([]);

  // adds event listeners when component mounts
  useEffect(() => {
    // connects to background.js
    let port = chrome.runtime.connect({ name: 'devtools-panel' });

    // listents for messages from npm package
    port.onMessage.addListener((message) => {
      if (message.type === 'event') {
        setQueryEvents((queryEvents) => [...queryEvents, message.event]);
      }

      if (message.type === 'metric') {
        console.log(message.metric);
      }
    });

    // reloads DevTool panel
    // need to define exact function for the listener to be removed in return
    const windowReloaded = () => {
      window.location.reload();
    };

    // event listner triggered when user navigates to new tab / reloads page
    chrome.devtools.network.onNavigated.addListener(windowReloaded);

    // cleanup 2 listeners on component dismount
    return () => {
      port.disconnect();
      chrome.devtools.network.onNavigated.removeListener(windowReloaded);
    };
  }, []);

  // updates state for selected queries
  const handleSelectionChange = (queries: string[]) => {
    setSelectedQueries(queries);
    saveSelectedQueryKeys(queries);
  };

  return (
    <Container
      maxWidth={false}
      sx={{
        height: '100vh',
        width: '100%',
        display: 'flex',
        flexDirection: 'column',
      }}
    >
      <ParentTab
        // queryOptions={queryOptions}
        queryEvents={queryEvents}
        selectedQueries={selectedQueries}
        handleSelectionChange={handleSelectionChange}
      />
    </Container>
  );
}

export default App;
