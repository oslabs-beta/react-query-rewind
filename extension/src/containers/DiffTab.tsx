import Typography from '@mui/material/Typography';
import FormGroup from '@mui/material/FormGroup';
import FormControlLabel from '@mui/material/FormControlLabel';
import Switch from '@mui/material/Switch';
import React, { useState } from 'react';

import JsonDiff from '../components/JsonDiff';

import { DataTabProps } from '../types';

const DiffTab = ({ queryDisplay, currentIndex }: DataTabProps) => {
  // state to determine if unchanged are hidden or closed
  const [isHidden, setIsHidden] = useState(false);

  // function to hide/show unchanged data
  const toggleChangedProperties = () => {
    setIsHidden(!isHidden);
  };

  return (
    <>
      <FormControlLabel
        sx={{ color: 'primary.main', marginLeft: 0 }}
        control={
          <Switch checked={isHidden} onChange={toggleChangedProperties} />
        }
        label={`${isHidden ? 'Show' : 'Hide'} Unchanged Properties `}
      />
      {queryDisplay.length > 0 && queryDisplay[currentIndex] && (
        <div className='data'>
          {queryDisplay[currentIndex].map((queryState, i) => (
            <>
              <Typography variant='h6' sx={{ color: 'secondary.main' }}>
                {queryState.queryKey}
              </Typography>
              <JsonDiff
                key={queryState.queryKey}
                queryKey={queryState.queryKey}
                currentJson={queryState.queryData}
                isHidden={isHidden}
                oldJson={
                  currentIndex > 1 && queryState.queryKey
                    ? queryDisplay[currentIndex - 1].find(
                        (obj) => obj.queryKey === queryState.queryKey
                      )?.queryData
                    : null
                }
              />
            </>
          ))}
        </div>
      )}
    </>
  );
};

export default DiffTab;
