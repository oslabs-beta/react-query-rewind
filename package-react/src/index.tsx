import React, { useEffect, useRef, useState } from 'react';
import Subscription from './Subscription';
import TimeTravel from './TimeTravel';
import ErrorBoundary from './ErrorBoundary';

function ReactQueryRewind() {
  const [timeTravel, setTimeTravel] = useState(false);
  const contentConnectedRef = useRef(false);
  const contentMessageQueueRef = useRef<any[]>([]);

  const sendContentMessageQueue = () => {
    contentMessageQueueRef.current.forEach(sendEvent);
    contentMessageQueueRef.current = [];
  };

  const handleMessages = (message: any) => {
    if (contentConnectedRef.current) {
      sendEvent(message);
    } else {
      contentMessageQueueRef.current.push(message);
    }
  };

  const sendEvent = (event: any) => {
    window.postMessage(
      { framework: 'react', type: 'event', payload: event },
      '*'
    );
  };

  const handleContentMessages = (message: MessageEvent) => {
    if (message.data?.type === 'content-script-ready') {
      // console.log('APP: Content.ts connected');
      contentConnectedRef.current = true;
      window.postMessage({ type: 'app-connected' }, '*');
      sendContentMessageQueue();
    }

    if (message.data?.type === 'time-travel') {
      setTimeTravel(message.data.payload);
    }
  };

  useEffect(() => {
    // console.log('APP: Mounting Event Listeners');
    window.addEventListener('message', handleContentMessages);

    return () => {
      window.removeEventListener('message', handleContentMessages);
    };
  }, []);

  return (
    <ErrorBoundary>
      {timeTravel ? (
        <TimeTravel />
      ) : (
        <Subscription handleMessages={handleMessages} />
      )}
    </ErrorBoundary>
  );
}

export default ReactQueryRewind;
