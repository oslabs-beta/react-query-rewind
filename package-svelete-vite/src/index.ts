// Import your Svelte components
import SvelteQueryRewind from './SvelteQueryRewind.svelte';

// Export your Svelte components
export { SvelteQueryRewind };
