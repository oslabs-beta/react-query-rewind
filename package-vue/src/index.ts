import VueQueryRewind from './VueQueryRewind.vue';

export default {
  install: (app: any) => {
    app.component('VueQueryRewind', VueQueryRewind);
  },
};
