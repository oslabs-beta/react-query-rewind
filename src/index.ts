// entry point that exports the useReactQueryRewind hook
