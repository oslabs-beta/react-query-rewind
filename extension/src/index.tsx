import React from 'react';
import ReactDOM from 'react-dom/client';
import App from './App';
import theme from './theme';
import { ThemeProvider, CssBaseline } from '@mui/material';

const root = document.createElement('div');
root.className = 'container';
document.body.appendChild(root);
const rootDiv = ReactDOM.createRoot(root);
rootDiv.render(
  <React.StrictMode>
    <ThemeProvider theme={theme}>
      <CssBaseline />
      <App />
    </ThemeProvider>
  </React.StrictMode>
);
