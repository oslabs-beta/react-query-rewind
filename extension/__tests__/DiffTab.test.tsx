import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import DiffTab from '../src/containers/DiffTab';
import '@testing-library/jest-dom';
import { QueryDisplay } from '../src/types';
import exp from 'constants';

describe('DiffTab Component', () => {
  const queryDisplayMock: QueryDisplay[][] = [
    [
      {
        queryKey: 'Query 1',
        queryData: { some: 'data' }
      },
      {
        queryKey: 'Query 2',
        queryData: { more: 'data' }
      }
    ],
    [
      {
        queryKey: 'Query 1',
        queryData: { some: 'different data' }
      }
    ]
  ];

  it('renders correctly with initial state', () => {
    const { asFragment } = render(<DiffTab queryDisplay={queryDisplayMock} currentIndex={0} />);
    expect(screen.getByLabelText(/Hide Unchanged Properties/i)).toBeInTheDocument();
    expect(screen.getByText(/Query 1/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
    });

  it('toggles isHidden state on switch toggle', () => {
    const { asFragment } = render(<DiffTab queryDisplay={queryDisplayMock} currentIndex={0} />);
    fireEvent.click(screen.getByRole('checkbox')); // Clicks the switch
    expect(screen.getByLabelText(/Show Unchanged Properties/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders JsonDiff component with correct props based on isHidden state', () => {
    const { rerender, asFragment } = render(<DiffTab queryDisplay={queryDisplayMock} currentIndex={1} />);
    let switchControl = screen.getByRole('checkbox');

    // Initially, isHidden should be false, so "Hide" should be in the label
    expect(switchControl).not.toBeChecked();
    expect(screen.getByText(/Hide Unchanged Properties/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();

    // Toggle the switch to change isHidden state
    fireEvent.click(switchControl);
    // Rerendering is necessary only if the component's response to state changes cannot be detected without it
    rerender(<DiffTab queryDisplay={queryDisplayMock} currentIndex={1} />);

    // Now, isHidden should be true, so "Show" should be in the label
    expect(screen.getByLabelText(/Show Unchanged Properties/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
