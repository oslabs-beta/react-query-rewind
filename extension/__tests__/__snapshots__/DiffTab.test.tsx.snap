// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiffTab Component renders JsonDiff component with correct props based on isHidden state 1`] = `
<DocumentFragment>
  <label
    class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-1yyhvac-MuiFormControlLabel-root"
  >
    <span
      class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
    >
      <span
        class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary css-5ryogn-MuiButtonBase-root-MuiSwitch-switchBase"
      >
        <input
          class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
          type="checkbox"
        />
        <span
          class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
        />
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
      />
    </span>
    <span
      class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
    >
      Hide Unchanged Properties 
    </span>
  </label>
  <div
    class="data"
  >
    <h6
      class="MuiTypography-root MuiTypography-h6 css-6zgmsr-MuiTypography-root"
    >
      Query 1
    </h6>
    <div
      class="json-diff-container "
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      >
        <div>
          <div
            class="jsondiffpatch-delta jsondiffpatch-modified"
          >
            <div
              class="jsondiffpatch-value jsondiffpatch-left-value"
            >
              <pre>
                null
              </pre>
            </div>
            <div
              class="jsondiffpatch-value jsondiffpatch-right-value"
            >
              <pre>
                {
  "some": "different data"
}
              </pre>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`DiffTab Component renders JsonDiff component with correct props based on isHidden state 2`] = `
<DocumentFragment>
  <label
    class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-1yyhvac-MuiFormControlLabel-root"
  >
    <span
      class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
    >
      <span
        class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked Mui-checked css-5ryogn-MuiButtonBase-root-MuiSwitch-switchBase"
      >
        <input
          class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
          type="checkbox"
        />
        <span
          class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
        />
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
      />
    </span>
    <span
      class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
    >
      Show Unchanged Properties 
    </span>
  </label>
  <div
    class="data"
  >
    <h6
      class="MuiTypography-root MuiTypography-h6 css-6zgmsr-MuiTypography-root"
    >
      Query 1
    </h6>
    <div
      class="json-diff-container jsondiffpatch-unchanged-hidden"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      >
        <div>
          <div
            class="jsondiffpatch-delta jsondiffpatch-modified"
          >
            <div
              class="jsondiffpatch-value jsondiffpatch-left-value"
            >
              <pre>
                null
              </pre>
            </div>
            <div
              class="jsondiffpatch-value jsondiffpatch-right-value"
            >
              <pre>
                {
  "some": "different data"
}
              </pre>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`DiffTab Component renders correctly with initial state 1`] = `
<DocumentFragment>
  <label
    class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-1yyhvac-MuiFormControlLabel-root"
  >
    <span
      class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
    >
      <span
        class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary css-5ryogn-MuiButtonBase-root-MuiSwitch-switchBase"
      >
        <input
          class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
          type="checkbox"
        />
        <span
          class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
        />
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
      />
    </span>
    <span
      class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
    >
      Hide Unchanged Properties 
    </span>
  </label>
  <div
    class="data"
  >
    <h6
      class="MuiTypography-root MuiTypography-h6 css-6zgmsr-MuiTypography-root"
    >
      Query 1
    </h6>
    <div
      class="json-diff-container "
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      >
        <div>
          <div
            class="jsondiffpatch-delta jsondiffpatch-modified"
          >
            <div
              class="jsondiffpatch-value jsondiffpatch-left-value"
            >
              <pre>
                null
              </pre>
            </div>
            <div
              class="jsondiffpatch-value jsondiffpatch-right-value"
            >
              <pre>
                {
  "some": "data"
}
              </pre>
            </div>
          </div>
        </div>
      </div>
    </div>
    <h6
      class="MuiTypography-root MuiTypography-h6 css-6zgmsr-MuiTypography-root"
    >
      Query 2
    </h6>
    <div
      class="json-diff-container "
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      >
        <div>
          <div
            class="jsondiffpatch-delta jsondiffpatch-modified"
          >
            <div
              class="jsondiffpatch-value jsondiffpatch-left-value"
            >
              <pre>
                null
              </pre>
            </div>
            <div
              class="jsondiffpatch-value jsondiffpatch-right-value"
            >
              <pre>
                {
  "more": "data"
}
              </pre>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`DiffTab Component toggles isHidden state on switch toggle 1`] = `
<DocumentFragment>
  <label
    class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-1yyhvac-MuiFormControlLabel-root"
  >
    <span
      class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
    >
      <span
        class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked Mui-checked css-5ryogn-MuiButtonBase-root-MuiSwitch-switchBase"
      >
        <input
          class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
          type="checkbox"
        />
        <span
          class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
        />
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
      />
    </span>
    <span
      class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
    >
      Show Unchanged Properties 
    </span>
  </label>
  <div
    class="data"
  >
    <h6
      class="MuiTypography-root MuiTypography-h6 css-6zgmsr-MuiTypography-root"
    >
      Query 1
    </h6>
    <div
      class="json-diff-container jsondiffpatch-unchanged-hidden"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      >
        <div>
          <div
            class="jsondiffpatch-delta jsondiffpatch-modified"
          >
            <div
              class="jsondiffpatch-value jsondiffpatch-left-value"
            >
              <pre>
                null
              </pre>
            </div>
            <div
              class="jsondiffpatch-value jsondiffpatch-right-value"
            >
              <pre>
                {
  "some": "data"
}
              </pre>
            </div>
          </div>
        </div>
      </div>
    </div>
    <h6
      class="MuiTypography-root MuiTypography-h6 css-6zgmsr-MuiTypography-root"
    >
      Query 2
    </h6>
    <div
      class="json-diff-container jsondiffpatch-unchanged-hidden"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      >
        <div>
          <div
            class="jsondiffpatch-delta jsondiffpatch-modified"
          >
            <div
              class="jsondiffpatch-value jsondiffpatch-left-value"
            >
              <pre>
                null
              </pre>
            </div>
            <div
              class="jsondiffpatch-value jsondiffpatch-right-value"
            >
              <pre>
                {
  "more": "data"
}
              </pre>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
